axlearn/common/checkpointer.py

```diff
@@ -922,3 +953,220 @@ def validate_and_restore(*, step: int, ckpt_dir: str):
+        )
+        tf_ckpt_handler, self._tf_ckpt_args = self._wrap_handler(
+            save_fn=lambda dir, args: save_tf_savables(args.item, dir=dir),
+            restore_fn=lambda dir, args: restore_tf_savables(args.item, dir=dir),
+        )
+
+        self._name_format = ocp.step.standard_name_format(
+            step_prefix=_STEP_PREFIX,
+            step_format_fixed_length=_STEP_NUM_DIGITS,
+        )
+        self._manager = ocp.CheckpointManager(
+            directory=cfg.dir,
+            options=ocp.CheckpointManagerOptions(
+                create=True,
+                max_to_keep=cfg.keep_last_n,
+                enable_async_checkpointing=True,
+                step_name_format=self._name_format,
+                should_save_fn=save_fn_with_summaries,
+            ),
+            item_handlers={
+                "index": index_handler(),
+                "tf_ckpt_map": tf_ckpt_handler(),
+                "state": ocp.StandardCheckpointHandler(save_concurrent_gb=cfg.save_concurrent_gb),
+            },
+        )
+
+    def _get_spec(self, *, step: int, state: NestedTensor) -> Nested[Any]:
+        spec = {"index": [("step", step)], "tf_ckpt_map": {}}
+        for path, value in utils.flatten_items(state):
+            if isinstance(value, (Tensor, TensorSpec)):
+                dtype = getattr(value.dtype, "dtype", value.dtype)
+                spec["index"].append(
+                    (path, {"dtype": str(dtype), "shape": str(tuple(value.shape))})
+                )
+            elif isinstance(value, tf.data.Iterator):
+                spec["index"].append((path, str(type(value))))
+                spec["tf_ckpt_map"][path] = value
+            else:
+                spec["index"].append((path, value))
+        return spec
+
+    # pylint: disable-next=redefined-builtin
+    def ckpt_dir(self, step: int, dir: Optional[str] = None) -> str:
+        """Obtains the checkpoint dir for the given step."""
+        import orbax.checkpoint as ocp
+
+        if dir is None:
+            dir = self._manager.directory
+        return str(ocp.step.build_step_path(dir, self._name_format, step))
+
+    def save(
+        self, *, step: int, state: Nested[Tensor], evaler_summaries: Optional[Dict[str, Any]] = None
+    ):
+        """See `BaseCheckpointer.save` for details.
+
+        Checkpoint saving is handled by `orbax` checkpoint manager.
+        """
+        import orbax.checkpoint as ocp
+
+        spec = self._get_spec(step=step, state=state)
+        self._eval_summaries_sem.acquire()  # pylint: disable=consider-using-with
+        assert not self._eval_summaries
+        self._eval_summaries.update(evaler_summaries or {})
+        self._manager.save(
+            step=step,
+            # The input iterator is saved as part of `save_tf_savables`.
+            args=ocp.args.Composite(
+                index=self._index_args(spec["index"]),
+                tf_ckpt_map=self._tf_ckpt_args(spec["tf_ckpt_map"]),
```

Comment:
What are `index` and `tf_ckpt_args`? Does orbax provide similar functionalities, eg. checking tree structure (which I assume is what the index file is for)?

Reply:
Good question! `index` is not just for validating structure but also restoring from tensor specs. `tf_ckpt_args` is for tf saveables like tf data iterators. LMK if orbax has built-in support for these.

Reply:
@cpgaffney1 Do you mind answering if orbax has built-in support for either `index` or `tf_ckpt_args` functionalities here?

Reply:
Sorry for the delay -- I've changed the internal impl slightly to use the json handler. Will push an update soon.